### ISSUE TITLE
wrong vendorid/devid with the right chip?

### ISSUE THREAD

**Issue by pclinuxer** [markusj/upd72020x-load#16]
Hello:
First of all, thanks for this utility.
I've been needing to get this card to work for the longest time.

I have come across a problem I don't know how to deal with.

My card is properly reported in lspsci:

[root@devuan work] # lspci 04:00.0 USB controller: Renesas Technology Corp. uPD720201 USB 3.0 Host Controller (rev 03)

ie: lists vendor as being Renesas - lists chip as being uPD720201

I got my hands on a FW file from this site:
https://raw.githubusercontent.com/chunkeey/renesas-fw/master/USB3-201-202-FW-20120615.zip

The file comes with a release note:

    D720201 & D720202 Design Resources Release
    Release Note June 15.2012
    Renesas Electronics D720201 & uPD720202
    USB3.0 Host Controller Firmware
    Version : 2.0.1.3 :June 15th, 2012
    Copyright (C) 2011-2012 Renesas Electronics Corporation All Rights Reserved

    Note : This firmware is for the following devices.
    - uPD720201 ES 2.0 sample whose revision ID (in the PCI Configuration
    Register) is 2h.
    - uPD720201 ES 2.1 sample & CS sample & Mass product whose revision
    ID (in the PCI Configuration Register) is 3h.
    - uPD720202 ES 2.0 sample & CS sample & Mass product whose revision
    ID (in the PCI Configuration Register) is 2h.

So it would seem (?) that what lspci is reporting would be the "uPD720201 Mass product whose Revision ID is 3h" device listed above.

But when I run the loader I get this:

[root@devuan work]# ./upd72020x-load -u -b 0x04 -d 0x00 -f -i 0x00 K2013080.mem Doing the upload bus = 4 dev = 0 fct = 0 fname = (null) ERROR: wrong vendorid/devid. Expected an UPD720201 or UPD720202 chip and this is not one! reported vendorid/devid: 1912:0014 ======> FAILED

It turns out that the vendorid/devid: 1912:0014 reported is (or is simply *reported* as being) the right one.

Here's the first hit for "vendorid/devid: 1912:0014" from a well known/ubiquitous search engine:

    Vendor Devices
    Type PCI
    Vendor ID 1912
    Vendor Name Renesas Technology Corp.
    Device ID 0014
    Device Name uPD720201 USB 3.0 Host Controller

So ...
Where's the discrepancy?
The firmware seems to be a correct match for the card but ...

Edit:
I saw this post https://github.com/markusj/upd72020x-load/issues/8 and unplugged a hub that was there, but the error subsists.

Edit: for completeness' sake, here is the output of  `lspci -nn`
``
04:00.0 USB controller [0c03]: Renesas Technology Corp. uPD720201 USB 3.0 Host Controller [1912:0014] (rev 03)
``
And there's no output in `dmesg` when this happens.

Edit 2:

If I try to save the original firmware I also get an error:
``
[root@devuan work]#   ./upd72020x-load -r -b 0x04 -d 0x00 -f 0x00 -s -o original.mem
Doing the reading
bus = 4 
dev = 0 
fct = 0 
fname = (null) 
ERROR: wrong vendorid/devid. Expected an UPD720201 or UPD720202 chip and this is not one!
       reported vendorid/devid: 1912:0014 
 ======> FAILED
[root@devuan work]# 
``
Any help you could give me would be greatly appreciated.

Thanks in advance.

**Comment by rtreffer** [comment]
I've noticed what I think is a bug, the `is_x1` is not necessarily initialized to `true`, could you try
https://github.com/markusj/upd72020x-load/pull/14/files#diff-1d3077c6fa70cefdca744d022ce66f2e8654077fdd6d3163afe32911db82100cR527 ?

**Comment by pclinuxer** [comment]
Hello:
>
> I've noticed what I think is a bug ...
> ... could you try ...
My knowledge of all this is rather limited, so please bear with me. 
Is that all the code I need to compile anotehr version to test?
ie: make upd72020x-load.c

Thanks in advance,

**Comment by rtreffer** [comment]
So yes in that file line 527 seems wrong. It currently reads `is_x1, is_x2
= true;` - this does not work in C if I recall correctly. `is_x1` is
undefined and `is_x2` is true (until checked).

So if you compiled the project in the past edit that line to read `is_x1 =
true, is_x2 = true;`, then compile as usual (`make` is enough - the task of
make is to figure out what changed an needs to be recompiled)

Then use the new binary to upload.

pclinuxer ***@***.***> schrieb am Mo., 22. März 2021, 18:27:

**Comment by pclinuxer** [comment]
Hello:

Right.
I have modified the line (526) to read like this:

` bool is_x1 = true, is_x2 = true;`   <--- now I have to remember that this is not v0.1 anymore.

The first this was to get a copy of the original firmware:

``
[root@devuan work]# ./upd72020x-load -r -b 0x04 -d 0x00 -f 0x00 -s -o original.mem
Doing the reading
bus = 4 
dev = 0 
fct = 0 
fname = (null) 
Found an UPD720201 chipset
got firmware version: 201809
EEPROM installed
got rom_info: 19d207f
got rom_config: 700
setting rom_config: 700
ERROR: cant open file (null)
 ======> FAILED
[root@devuan work]# 
``
Now the printout is more complete.
ie: 

Found an UPD720201 chipset    <- finds the chipset
got firmware version: 201809     <- gets the FW version
EEPROM installed                      <- confirms EEPROM is installed
got rom_info: 19d207f                <
got rom_config: 700                    <
setting rom_config: 700              <

But end with an error message and does not save the FW.
ie: original.mem 

Next was to try and see if it will write one of the firmware files I have:

``
[root@devuan work]# ./upd72020x-load -w -b 0x04 -d 0x00 -f 0x00 -i K2013080.mem
Doing the writing
bus = 4 
dev = 0 
fct = 0 
fname = K2013080.mem 
Found an UPD720201 chipset
got firmware version: 201809
EEPROM installed
got rom_info: 19d207f
got rom_config: 700
setting rom_config: 700
STATUS: enabling EEPROM write
STATUS: performing EEPROM write
STATUS: finishing EEPROM write
STATUS: confirming EEPROM write
 ======> PASSED
[root@devuan work]# 
``
I seems it did.
But the FW version has seems to be the same one.
And plugging in the drive makes no difference, still not detected.

Comments?

Thanks in advance,

**Comment by rtreffer** [comment]
Flashing (writing, `-w`) the firmware is as far as I know not the thing you want to do. You would want to upload (`-u`). Writing the firmware to eeprom has no effect unless a reload from flash is triggered (according to the datasheet). It might cause issues though....

I have done the same mistake, although a bit more intentional (YOLO ftw).... I have documented how to flash / recover the chip in
https://github.com/markusj/upd72020x-load/issues/15 and https://github.com/geerlingguy/raspberry-pi-pcie-devices/issues/103

Long story short: you need to prepare the firmware for an upload.
Btw, 201809 stands for `2.0.1.8` and you are trying to flash a `2.0.1.3`. So you are trying a downgrade at the moment.

Your flash chip seems to be a `Chingis / Pm25LD512C2` with 16kb block size which is supported by flashrom, so you might be able to recover everything (flashing through upd72020x-load updates the second copy of the firmware - unless an erase is triggered first).

**Comment by rtreffer** [comment]
Ah, and don't use `-s`.... Try `./upd72020x-load -r -b 0x04 -d 0x00 -f 0x00 -o original.mem` - the tool dumps the flag based config first and `fname = (null)` is clearly wrong (and results in `ERROR: cant open file (null)`).

**Comment by pclinuxer** [comment]
Hello:
>
> Flashing (writing, -w) the firmware ... 
> ... not the thing you want to do.
> You would want to upload (-u). 
Right.
I had thought about it but eeventually forgot.
Apparently I have *not* bricked the card.  8^D

>
> ... don't use -s.... Try `./upd72020x-load -r -b 0x04 -d 0x00 -f 0x00 -o original.mem`
Got it, thanks.

[code]
[root@devuan work]# ./upd72020x-load -r -b 0x04 -d 0x00 -f 0x00 -o original.mem
Doing the reading
bus = 4 
dev = 0 
fct = 0 
fname = original.mem 
Found an UPD720201 chipset
got firmware version: 201809
EEPROM installed
got rom_info: 19d207f
got rom_config: 700
setting rom_config: 700
 ======> PASSED
[root@devuan work]# 
[/code]

And there it is.

[code]
[root@devuan work]# ls | grep .mem
K2013080.mem
K2020090.mem
K2024.mem
original.mem
[root@devuan work]# 
[/code]

So if `K2013080.mem` would be a downgrade `K2020090.mem` would not because the original one is K2013080.mem.
I'll see how that goes.

Thanks so much for your help.
Best,

**Comment by rtreffer** [comment]
Version numbers are just nibbles (half bytes) as far as I can tell. I haven't fully decoded the meaning of these yet so it might be interesting what your mem file collection has there.
You can find the firmware version at offset 0x12 (unverified - my wild guess).
E.g. `K2013080` has `0x08 0x00 0x13 0x20`. Which is often "named" `2.0.1.3`. I think this is loaded as uint32 little endian, it gets loaded as 0x20130008.

I had good results with 202609 which should be ` K2020090.mem`. I have never seen the `K2024.mem` though.

**Comment by pclinuxer** [comment]
Hello:
>
> ... might be interesting what your mem file collection has there.
>
Be glad to send them to you or upload them somewhere.
Let me know.

Best,

**Comment by rtreffer** [comment]
No need for an upload, but you could dump the 4 bytes from 18-22 and post it to (dis)prove the firmware version position:
```
for i in *.mem; do
  echo $i $( head -c 22 $i |tail -c 4| hexdump -C |head -n1 )
done
```

**Comment by pclinuxer** [comment]
Hello:

Here it is.

[code]
groucho@devuan:~/Downloads/renesas/work$ for i in *.mem; do   echo $i $( head -c 22 $i |tail -c 4| hexdump -C |head -n1 ); done

K2013080.mem 00000000 08 00 13 20 |... |
K2024.mem 00000000 08 00 24 20 |..$ |
original.mem 00000000 10 00 11 00 |....|

groucho@devuan:~/Downloads/renesas/work$ 
[/code]

**Comment by rtreffer** [comment]
Ah! `10 00 11 00` is part of the vendor specific configuration...

Does `hexdump -C original.mem` start like this?
```
00000000  28 00 00 ff 01 ff 02 ff  03 ff 0c 00 0d 00 0e 00  |(...............|
00000010  0f 00 10 00 11 00 12 00  13 00 14 00 15 00 16 55  |...............U|
00000020  17 00 18 00 19 00 1a 01  1b 05 e7 ef aa 55        |.............   |
```
(`aa 55` is the beginning of the actual firmware, the first 2 bytes (`0x28 0x00 == 0x0028 == 40`) indicate the length of the configuration without crc: `0x0028` - so the firmware starts at `0x28 + 2 + 2 = 0x2c`)

To flash with `pd72020x-load` you would need a 16kb file (the block size of the flash chip) consisting of the vendor specific configuration followed by the firmware (e.g. `K2024.mem`) followed by `0xff` until the the end of file. Then write that file with `pd72020x-load -w` and reboot or better power off and on.
That's highly experimental though, but could remove the need to ever touch this tool again.

**Comment by pclinuxer** [comment]
Hello:
>
> Ah! 10 00 11 00 is part of the vendor specific configuration...
>
So you've cracked it.  8^D
>
> Does hexdump -C original.mem start like this?
>
Here are the first five lines of the *.mem files I have:
[code]
groucho@devuan:~/Downloads/renesas/work$ hexdump -C original.mem
00000000  28 00 00 ff 01 ff 02 ff  03 ff 0c 00 0d 00 0e 00  |(...............|
00000010  0f 00 10 00 11 00 12 00  13 00 14 00 15 00 16 55  |...............U|
00000020  17 00 18 00 19 00 1a 01  1b 05 e7 ef aa 55 17 f8  |.............U..|
00000030  14 00 f6 00 0c 01 4c 00  5c 01 20 2e 80 2f 08 00  |......L.\. ../..|
00000040  18 20 5c 1a 63 ff 01 00  80 ff 1e 00 23 ff 01 00  |. \.c.......#...|
--- snip ---
[/code]
[code]
groucho@devuan:~/Downloads/renesas/work$ hexdump -C K2024.mem
00000000  aa 55 a6 7d 14 00 f6 00  0c 01 4c 00 5c 01 fc 2f  |.U.}......L.\../|
00000010  5c 31 08 00 24 20 5c 1a  63 ff 01 00 80 ff 1e 00  |\1..$ \.c.......|
00000020  23 ff 01 00 44 1a 7f 00  5c 1a 63 ff 01 00 80 ff  |#...D...\.c.....|
00000030  14 03 23 ff 01 00 44 1a  7f 00 80 07 21 00 25 06  |..#...D.....!.%.|
00000040  d4 aa 01 00 05 2e 00 40  05 2e 00 40 25 f6 00 80  |.......@...@%...|
[/code]
[code]
groucho@devuan:~/Downloads/renesas/work$ hexdump -C K2013080.mem
00000000  aa 55 c7 3f 14 00 f6 00  0c 01 54 00 64 01 94 2c  |.U.?......T.d..,|
00000010  fc 2d 08 00 13 20 5c 1a  63 ff 01 00 80 ff 1e 00  |.-... \.c.......|
00000020  23 ff 01 00 44 1a 7f 00  5c 1a 63 ff 01 00 80 ff  |#...D...\.c.....|
00000030  14 03 23 ff 01 00 44 1a  7f 00 80 07 21 00 25 06  |..#...D.....!.%.|
00000040  6c a7 01 00 05 2e 00 40  05 2e 00 40 25 f6 00 80  |l......@...@%...|
[/code]

All this is rather on the esoteric side of IT for me.

I'm a hardware guy: will take apart anything, given the adecquate instruments.
But this ...  

Off to look for a FW file.
All I get are MS based executables.

Best,

**Comment by rtreffer** [comment]
You can try this
```
echo -ne '\x28\x00\x00\xff\x01\xff\x02\xff\x03\xff\x0c\x00\x0d\x00\x0e\x00\x0f\x00\x10\x00\x11\x00\x12\x00\x13\x00\x14\x00\x15\x00\x16\x55\x17\x00\x18\x00\x19\x00\x1a\x01\x1b\x05\xe7\xef'
  > vendor.cfg
cat vendor.cfg K2024.mem > new.flash
```

Then try uploading that file
```
./upd72020x-load ... -w -f new.flash
```
Then power off / power on and check if reading the FW shows a different version....

**Comment by pclinuxer** [comment]
Hello:
> Then power off / power on and check if reading the FW shows a different version....
>
Does unloading and reloading the pertinent module do the same thing?

ie: 
[code]
[root@devuan ~]# modprobe -r xhci_pci
[root@devuan ~]# modprobe xhci_pci
[/code]

Thanks for you input.

Best,

**Comment by rtreffer** [comment]
No, the card needs a reboot.

A PCI reset might help, but I have not tried that. There is a flash reload
command, but that's not implemented in this tool as far as I know....

A Reboot might do. A power off/on is safer.

I'll sign off for today. It's late....

pclinuxer ***@***.***> schrieb am Di., 23. März 2021, 00:10:

**Comment by pclinuxer** [comment]
Hello:

> No, the card needs a reboot.
Right.
Shutdown, turned power to the PS off, on again after a minute and rebooted.

I then saved the uploaded file as new2.mem:

[code]
[root@devuan work]# ./upd72020x-load -r -b 0x04 -d 0x00 -f 0x00 -o new2.mem
Doing the reading
bus = 4 
dev = 0 
fct = 0 
fname = new2.mem 
Found an UPD720201 chipset
got firmware version: 202409
EEPROM installed
got rom_info: 19d207f
got rom_config: 0
setting rom_config: 700
 ======> PASSED
[root@devuan work]# 
[/code]

I now have FW version 202409.
ie: got firmware version: 202409

The one in the card used to be 201809.
Just to see, I unloaded/reloaded the 

[code]
[root@devuan ~]# modprobe -r xhci_pci
[root@devuan ~]# modprobe -r xhci_pci
[/code]

This is the `dmesg`printout:
The cards reloads and seems to do it twice, at bus number 4 and bus number 5 (?)

[code]
[  145.224823] elogind-daemon[2224]: New session 1 of user groucho.
[  145.789627] fuse init (API version 7.27)

Unloaded xHCI module

[  915.611556] xhci_hcd 0000:04:00.0: remove, state 4
[  915.611568] usb usb5: USB disconnect, device number 1
[  915.611847] xhci_hcd 0000:04:00.0: USB bus 5 deregistered
[  915.611859] xhci_hcd 0000:04:00.0: remove, state 4
[  915.611865] usb usb4: USB disconnect, device number 1
[  915.611868] usb 4-3: USB disconnect, device number 2
[  915.627346] xhci_hcd 0000:04:00.0: USB bus 4 deregistered

Reloaded xHCI module
--->  At bus number 4
[  920.245318] xhci_hcd 0000:04:00.0: xHCI Host Controller
[  920.245327] xhci_hcd 0000:04:00.0: new USB bus registered, assigned bus number 4
[  920.250691] xhci_hcd 0000:04:00.0: hcc params 0x014051cf hci version 0x100 quirks 0x0000000100000410
[  920.250943] usb usb4: New USB device found, idVendor=1d6b, idProduct=0002, bcdDevice= 4.19
[  920.250946] usb usb4: New USB device strings: Mfr=3, Product=2, SerialNumber=1
--- >  At bus number 5
[  920.250948] usb usb4: Product: xHCI Host Controller
[  920.250950] usb usb4: Manufacturer: Linux 4.19.0-14-amd64 xhci-hcd
[  920.250951] usb usb4: SerialNumber: 0000:04:00.0
[  920.251066] hub 4-0:1.0: USB hub found
[  920.251079] hub 4-0:1.0: 4 ports detected
[  920.251274] xhci_hcd 0000:04:00.0: xHCI Host Controller
[  920.251278] xhci_hcd 0000:04:00.0: new USB bus registered, assigned bus number 5
[  920.251281] xhci_hcd 0000:04:00.0: Host supports USB 3.0 SuperSpeed
[  920.254283] usb usb5: We don't know the algorithms for LPM for this host, disabling LPM.
[  920.254303] usb usb5: New USB device found, idVendor=1d6b, idProduct=0003, bcdDevice= 4.19
[  920.254306] usb usb5: New USB device strings: Mfr=3, Product=2, SerialNumber=1
[  920.254307] usb usb5: Product: xHCI Host Controller
[  920.254309] usb usb5: Manufacturer: Linux 4.19.0-14-amd64 xhci-hcd
[  920.254311] usb usb5: SerialNumber: 0000:04:00.0
[  920.254412] hub 5-0:1.0: USB hub found
[  920.254426] hub 5-0:1.0: 4 ports detected
This is the Dell monitor hub.
[  920.583321] usb 4-3: new high-speed USB device number 2 using xhci_hcd
[  920.732417] usb 4-3: New USB device found, idVendor=0424, idProduct=2514, bcdDevice= b.b3
[  920.732421] usb 4-3: New USB device strings: Mfr=0, Product=0, SerialNumber=0
[  920.736460] hub 4-3:1.0: USB hub found
[  920.736787] hub 4-3:1.0: 4 ports detected
[root@devuan ~]# 

But the card still does not detect my USB3.0 external drive.
 
> I'll sign off for today. It's late.... 
>
Yes, I was wondering.
I'm at  GMT -03.00, you must be +01:00. 
Yes, it is late.

Thank you very much for your efforts.

Best,

**Comment by rtreffer** [comment]
So the symptoms of an old / missing firmware are that the card does not
start. You would see xhci fail on initialize.

The card should exist twice IIRC, once for USB2 and once for USB3.

This looks all pretty decent, except that 4.19 is a really old kernel.

Detecting other devices (hubs) means it works. Not a card FW issue. You
have working setup/card.

`lsusb -t` produces nice overviews.

The uPD720201 has quite some features, especially port power control and
overcurrent protection. That might be interesting for external drives.
And while I haven't tried it you might try to turn ports of the card on/off
with https://github.com/mvp/uhubctl

This depends on the electrical details of the card though.

pclinuxer ***@***.***> schrieb am Di., 23. März 2021, 01:01:

**Comment by pclinuxer** [comment]
Hello:

> The card should exist twice IIRC, once for USB2 and once for USB3.
>
Apparently it does.
In my case USB4 and USB5.

> ... 4.19 is a really old kernel.
>
That is Debian, stable with no bleeding edge attached.
I had my share of unstable in 20+ years of working with MS OSs.

I run Devuan Beowulf is just Debian Buster *without* systemd.
The 4.19.171-2 kernel is the current stable. 

That said, before uploading the new firmware I tested for the possibility of the cause of the problem being an older kernel with the latest version of Ubuntu fwts-live *.iso, which runs on a 5.x kernel.

The symproms were exactly the same as with my current setup.
ie: does not detect my USB3.0 external drive while and produces this line in `dmesg`: 

`[ 920.254283] usb usb5: We don't know the algorithms for LPM for this host, disabling LPM.`

But if the USB2 side of the card works properly, then it would be exclusively a USB3.0 problem. (?)

Two of the card's ports are in use, one by a printer and the other by the monitor's hub.
Both are properly detected.

If I plug the USB3.0 drive into the hub, the drive is detected and works, albeit at USB2.0 speeds.
If I also plug in an old external 40Gb USB2.0 ATA drive, it also works properly albeit with external power from another USB port like it always did, independently of being used with a box or a laptop.

From `dmesg`: 
[code]
`Printer`
1536.139279] usb 6-1: new high-speed USB device number 4 using ehci-pci
[ 1536.296149] usb 6-1: New USB device found, idVendor=04e8, idProduct=3321, bcdDevice= 1.00
[ 1536.296155] usb 6-1: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[ 1536.296158] usb 6-1: Product: M2020 Series
[ 1536.296161] usb 6-1: Manufacturer: Samsung Electronics Co., Ltd.
[ 1536.296163] usb 6-1: SerialNumber: 0731B8GJBB029TD
[ 1536.318107] usblp 6-1:1.0: usblp0: USB Bidirectional printer dev 4 if 0 alt 0 proto 2 vid 0x04E8 pid 0x3321
[ 1536.318135] usbcore: registered new interface driver usblp
`USB3.0 Drive`
[ 1945.779284] usb 4-3.1: new high-speed USB device number 3 using xhci_hcd
[ 1945.887039] usb 4-3.1: New USB device found, idVendor=043e, idProduct=70f5, bcdDevice= 6.00
[ 1945.887043] usb 4-3.1: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[ 1945.887046] usb 4-3.1: Product: LG External HDD
[ 1945.887049] usb 4-3.1: Manufacturer: LG Electronics Inc.
[ 1945.887051] usb 4-3.1: SerialNumber: A1204000000004C7
[ 1945.918261] usb-storage 4-3.1:1.0: USB Mass Storage device detected
[ 1945.918386] scsi host8: usb-storage 4-3.1:1.0
[ 1945.918512] usbcore: registered new interface driver usb-storage
[ 1949.831727] scsi 8:0:0:0: Direct-Access     Hitachi  HTS545050A7E     GG2O PQ: 0 ANSI: 2
[ 1949.832059] sd 8:0:0:0: Attached scsi generic sg6 type 0
[ 1949.836242] sd 8:0:0:0: [sdf] 976773164 512-byte logical blocks: (500 GB/466 GiB)
[ 1949.836767] sd 8:0:0:0: [sdf] Write Protect is off
[ 1949.836770] sd 8:0:0:0: [sdf] Mode Sense: 00 06 00 00
`USB2.0 Drive`
[ 2790.899310] usb 4-1: new full-speed USB device number 4 using xhci_hcd
[ 3295.103324] usb 4-1: new high-speed USB device number 5 using xhci_hcd
[ 3295.258517] usb 4-1: New USB device found, idVendor=05e3, idProduct=0702, bcdDevice= 0.02
[ 3295.258521] usb 4-1: New USB device strings: Mfr=2, Product=3, SerialNumber=0
[ 3295.258524] usb 4-1: Product: TS0MHDENC
[ 3295.258527] usb 4-1: Manufacturer: Transcend
[ 3295.264428] usb-storage 4-1:1.0: USB Mass Storage device detected
[ 3295.264606] usb-storage 4-1:1.0: Quirks match for vid 05e3 pid 0702: 520
[ 3295.264650] scsi host9: usb-storage 4-1:1.0
[ 3295.285941] usbcore: registered new interface driver uas
[ 3296.274177] scsi 9:0:0:0: Direct-Access     SAMSUNG  MP0402H          0811 PQ: 0 ANSI: 0
[ 3296.274480] sd 9:0:0:0: Attached scsi generic sg7 type 0
[ 3296.275033] sd 9:0:0:0: [sdg] 78242976 512-byte logical blocks: (40.1 GB/37.3 GiB)
[ 3296.277329] sd 9:0:0:0: [sdg] Test WP failed, assume Write Enabled
[ 3296.278867] sd 9:0:0:0: [sdg] Cache data unavailable
[ 3296.278872] sd 9:0:0:0: [sdg] Assuming drive cache: write through
[/code]  

> Detecting other devices (hubs) means it works. Not a card FW issue. You have working setup/card. 
>
I'm glad to know that.
Thank you.

> `lsusb -t` produces nice overviews.
>
Yes.

[code]
groucho@devuan:~$ lsusb -t
--- snip ---
/:  Bus 05.Port 1: Dev 1, Class=root_hub, Driver=xhci_hcd/4p, 5000M
/:  Bus 04.Port 1: Dev 1, Class=root_hub, Driver=xhci_hcd/4p, 480M
    |__ Port 1: Dev 5, If 0, Class=Mass Storage, Driver=usb-storage, 480M
    |__ Port 3: Dev 2, If 0, Class=Hub, Driver=hub/4p, 480M
        |__ Port 1: Dev 3, If 0, Class=Mass Storage, Driver=usb-storage, 480M
--- snip ---
groucho@devuan:~$ 
[code]

But if I move the USB3.0 drive to a port on the card it does not get detected.
It only gets detected through a USB2.0 hub.
Unfortunately, I don't have a USB3.0 hub to try.

> ... might try to turn ports of the card on/off with https://github.com/mvp/uhubctl
> ... depends on the electrical details of the card though.
The card, in it's 2-port version, is listed as being compatible:

`Renesas uPD720202 PCIe USB 3.0 host controller    2    3.0    2013`

I'll download it and see what it reports.
Maybe there is something revealing in the printout.

Thank you very much for taking the time to herlp me deal with this.
Much appreciated.

Best,

**Comment by rtreffer** [comment]
The thing is: many drives draw more current than the official USB spec
allows.

The uPD720201 supports over current protection and may turn ports
completely off if that triggers.

This would be in line with the symptoms and we'll within the spec....

Other than that check the cable, some don't have all lines going through

pclinuxer ***@***.***> schrieb am Di., 23. März 2021, 12:22:

**Comment by pclinuxer** [comment]
Hello:

I (think) I have been able to solve the riddle but not the problem.
On a whim, I connected another 500Gb USB3.0 drive I have, unused in a box.

It worked: same cable, same USB3.0 card.
Both drives are Toshiba but they are different models.

The one that works is model HXD7 USB 3.0 
The one that does not work is model HXD7 USB 3.0.

`dmesg`
[code]
[ 2420.760462] usb 2-1: new SuperSpeed Gen 1 USB device number 4 using xhci_hcd
[ 2420.787113] usb 2-1: New USB device found, idVendor=043e, idProduct=70f5, bcdDevice= 1.00
[ 2420.787117] usb 2-1: New USB device strings: Mfr=2, Product=3, SerialNumber=1
[ 2420.787120] usb 2-1: Product: LG External HDD
[ 2420.787123] usb 2-1: Manufacturer: LG Electronics Inc.
[ 2420.787125] usb 2-1: SerialNumber: A3110300000A
[ 2420.791771] usb-storage 2-1:1.0: USB Mass Storage device detected
[ 2420.791953] scsi host9: usb-storage 2-1:1.0
[ 2424.446142] scsi 9:0:0:0: Direct-Access              LG External HDD  AX00 PQ: 0 ANSI: 5
[ 2424.446523] sd 9:0:0:0: Attached scsi generic sg7 type 0
[ 2424.446645] sd 9:0:0:0: [sdg] 976773168 512-byte logical blocks: (500 GB/466 GiB)
[ 2424.447552] sd 9:0:0:0: [sdg] Write Protect is off
[ 2424.447555] sd 9:0:0:0: [sdg] Mode Sense: 23 00 00 00
[/code]

Both drives work properly, it's just that one of them does not do it at USB3.0 speeds.
Since both drives work with the same cable and the same USB.03 card, the problem has to be with whatever gets done between the card and the SATA/USB bridge.

Can't think of nothing else.

Model HXD7 has a ASM1053 SATA/USB controller chip.
[http://j5d2v7d7.stackpathcdn.com/wp-content/uploads/2015/10/ASM1051.gif] (url) photo from the web

Model HXD7 has a VIA Labs VL 701-04 
[https://www.via-labs.com/archive/images/editor/product/VL701_.jpg] (url) photo from tthe web

Plugged in the way we know work, uhubctl printout is clear: both are detected

[code]
groucho@devuan:~ $ sudo uhubctl
Current status for hub 2 [1d6b:0003 Linux 4.19.0-14-amd64 xhci-hcd xHCI Host Controller 0000:04:00.0, USB 3.00, 4 ports, ppps]
  Port 1: 0203 power 5gbps U0 enable connect [043e:70f5 LG Electronics Inc. LG External HDD A3110300000A]
  Port 2: 02a0 power 5gbps Rx.Detect
  Port 3: 02a0 power 5gbps Rx.Detect
  Port 4: 02a0 power 5gbps Rx.Detect
Current status for hub 1-3 [0424:2514, USB 2.00, 4 ports, ppps]
  Port 1: 0503 power highspeed enable connect [043e:70f5 LG Electronics Inc. LG External HDD A1204000000004C7]
  Port 2: 0100 power
  Port 3: 0100 power
  Port 4: 0100 power
Current status for hub 1 [1d6b:0002 Linux 4.19.0-14-amd64 xhci-hcd xHCI Host Controller 0000:04:00.0, USB 2.00, 4 ports, ppps]
  Port 1: 0100 power
  Port 2: 0100 power
  Port 3: 0503 power highspeed enable connect [0424:2514, USB 2.00, 4 ports, ppps]
  Port 4: 0100 power
groucho@devuan:~ $ 
[/code]

But when I switch them around, the one with the VIA SATA/USB bridge is no longer detected:

[code]
groucho@devuan:~ $ sudo uhubctl
Current status for hub 2 [1d6b:0003 Linux 4.19.0-14-amd64 xhci-hcd xHCI Host Controller 0000:04:00.0, USB 3.00, 4 ports, ppps]
  Port 1: 02a0 power 5gbps Rx.Detect
  Port 2: 02a0 power 5gbps Rx.Detect
  Port 3: 02a0 power 5gbps Rx.Detect
  Port 4: 02a0 power 5gbps Rx.Detect
Current status for hub 1-3 [0424:2514, USB 2.00, 4 ports, ppps]
  Port 1: 0503 power highspeed enable connect [043e:70f5 LG Electronics Inc. LG External HDD A3110300000A]
  Port 2: 0100 power
  Port 3: 0100 power
  Port 4: 0100 power
Current status for hub 1 [1d6b:0002 Linux 4.19.0-14-amd64 xhci-hcd xHCI Host Controller 0000:04:00.0, USB 2.00, 4 ports, ppps]
  Port 1: 0100 power
  Port 2: 0100 power
  Port 3: 0503 power highspeed enable connect [0424:2514, USB 2.00, 4 ports, ppps]
  Port 4: 0100 power
groucho@devuan:~ $ 
[/code]

That seems to be what it is all about.

If I had to make an (un)educated guess, I'd say that one SATA/USB bridge has a problem with the USB3.0 ports' Rx.Detect.
And if this were true, it may be a xHCI_PCI module problem/setting causing this.
But it is all over my head.

What do you think?

Thanks in advance.

Best,

**Comment by rtreffer** [comment]
No idea to be honest. I would write this off as "potentially out of spec,
so working sometimes".

Figuring out what is faulty is like a big try & error task and nothing I
can predict from here or really help with.... Power supply / requirements
are a likely cause, as well as signal integrity (was it 5GHz for USB3?).
Hard to tell, but that's where testing different cables / enclosures etc.
comes into play. You could also try to turn off all other ports. I have
seen one report that this can hap with cross talk to audio devices... Or
add an external _powered_ hub to make sure you are not running over the
power budget.

Anyway, it looks like you got all the hints and details to debug this, and
it is no longer an issue with this tool, so this can be closed, right?

pclinuxer ***@***.***> schrieb am Di., 23. März 2021, 17:27:

**Comment by pclinuxer** [comment]
Hello:
> No idea ...
> ... potentially out of spec, so working sometimes". 
>
Yes, that could be one option.

> ... nothing I can predict from here or really help with ...
> ... like you got all the hints and details to debug this,
Yes, you have been very helpful.
Learned quite bit.

> ... no longer an issue with this tool, so this can be closed, right?
Yes, by all means.

Once again, thanks a lot for your help.

Best,

**Comment by pclinuxer** [comment]
Hello:

Well ...
I deserve to be mocked to extintion.

Got these  USB drives long ago to do some installations that were eventually cancelled and got to keep them, have had very little use and are in pristine condition.

Each came with it's own OEM cable in the box and are, as far as I can see, `identical`.
I assumed that they are also `electrically` identical, no reason not to.
More so coming from the same OEM that sells these drives, both USB3.0. 

So ...
Cable `A` is the same as cable `B`, in `every` sense.
Right?
But you `did` mention cable more than once.

Well ...
Tl,dr: I switched the cables and now `both` drives are recognised.
[code]
groucho@devuan:~ $ sudo uhubctl
[sudo] password for groucho: 
Current status for hub 5 [1d6b:0003 Linux 4.19.0-14-amd64 xhci-hcd xHCI Host Controller 0000:04:00.0, USB 3.00, 4 ports, ppps]
  Port 1: 0203 power 5gbps U0 enable connect [043e:70f5 LG Electronics Inc. LG External HDD A3110300000A]
  Port 2: 0203 power 5gbps U0 enable connect [043e:70f5 LG Electronics Inc. LG External HDD A1204000000004C7]
  Port 3: 02a0 power 5gbps Rx.Detect
  Port 4: 02a0 power 5gbps Rx.Detect
Current status for hub 3-3 [0424:2514, USB 2.00, 4 ports, ppps]
  Port 1: 0100 power
  Port 2: 0100 power
  Port 3: 0100 power
  Port 4: 0100 power
Current status for hub 3 [1d6b:0002 Linux 4.19.0-14-amd64 xhci-hcd xHCI Host Controller 0000:04:00.0, USB 2.00, 4 ports, ppps]
  Port 1: 0100 power
  Port 2: 0100 power
  Port 3: 0507 power highspeed suspend enable connect [0424:2514, USB 2.00, 4 ports, ppps]
  Port 4: 0100 power
groucho@devuan:~ $ 
[/code]

For whatever reason, drive A will `only` work properly with cable A but drive B will work properly with cable A `or` cable B.

Unbelievable ...

Once again, thanks.

Best,

**Comment by markusj** [comment]
The wrong message has been fixed in #14 and everything appears to work for you now, so I close this issue for now.

@rtreffer: Thank you very much for the support you provided here!
